src/types/middleware.ts

```diff
@@ -0,0 +1,30 @@
+import { EffectScope } from 'vue'
+import { MaybePromise } from '@/types/utilities'
+import { Router } from '@/utilities'
+
+type Route = {
+  name: string,
+  params: Record<string, unknown>,
+  query: unknown,
+  hash: string,
+}
+
+interface RouteState {
+
+}
+
+interface RouteRejection {
+  NotFound: unknown,
+}
+
+type RouteReject = (type?: keyof RouteRejection) => void
+
+type MiddlewareExtras = {
+  from: Route | null,
+  state: RouteState,
+  reject: RouteReject,
+  scope: EffectScope,
+  router: Omit<Router<[]>, 'routes'> & { routes: Record<string, any> },
```

Comment:
We'll likely want a better type for this at some point. But for now this works. Will revisit once we spec out what `Router` is more.